src/common/tusb_fifo.c

```diff
@@ -95,9 +95,9 @@ static inline uint16_t _ff_mod(uint16_t idx, uint16_t depth)
 
 // Intended to be used to read from hardware USB FIFO in e.g. STM32 where all data is read from a constant address
 // Code adapted from dcd_synopsis.c
-static void _tu_fifo_read_from_const_src_ptr(void * dst, const void * src, uint16_t len)
+static void _tu_fifo_read_from_const_src_ptr(void * __restrict dst, const void * __restrict src, uint16_t len)
 {
-  uint8_t * dst_u8 = (uint8_t *)dst;
+  uint8_t CFG_TUSB_MEM_ALIGN * dst_u8 = (uint8_t *)dst;
```

Comment:
DMA is already covered by the functions
```
// Pointer modifications intended to be used in combinations with DMAs.
// USE WITH CARE - NO SAFTY CHECKS CONDUCTED HERE! NOT MUTEX PROTECTED!
void     tu_fifo_advance_write_pointer  (tu_fifo_t *f, uint16_t n);
void     tu_fifo_backward_write_pointer (tu_fifo_t *f, uint16_t n);
void     tu_fifo_advance_read_pointer   (tu_fifo_t *f, uint16_t n);
void     tu_fifo_backward_read_pointer  (tu_fifo_t *f, uint16_t n);

// If you want to read/write from/to the FIFO by use of a DMA, you may need to conduct two copies to handle a possible wrapping part
// This functions deliver a pointer to start reading/writing from/to and a valid linear length along which no wrap occurs.
// In case not all of your data is available within one read/write, update the read/write pointer by
// tu_fifo_advance_read_pointer()/tu_fifo_advance_write_pointer and conduct a second read/write operation
uint16_t tu_fifo_get_linear_read_info   (tu_fifo_t *f, uint16_t offset, void **ptr, uint16_t n);
uint16_t tu_fifo_get_linear_write_info  (tu_fifo_t *f, uint16_t offset, void **ptr, uint16_t n);
```

It depends on the DMA mode how you set up everything. E.g. if DMA is in circular mode writing into the fifo, the only thing to do is to update the write pointer within any suitable interrupt e.g. `tud_audio_tx_done_pre_load_cb()`.

If you want to copy out of the fifo, usually DMAs do not support circular mode (at least for stm32, there the fifo would copy infinitely). So in linear mode, you need to cover two copies since a wrap may occur. Therefore you use `tu_fifo_get_linear_read_info()` to determine the number of bytes the DMA can copie in one rush and then query again if a second copy for the wrapped part is required. If you have a more sophisticated DMA available (one that can do circular copies from memory) you may do otherwise.

Reply:
I though `memcpy()` already takes care of how the payload needs to be copied byte or word-wise?

Reply:
> 
> 
> > Yes alternate settings are implemented! But this was included from beginning on?! So you can define alternate interface descriptors 1,2,3 etc. and the host selects those. There you can define your desired sampling ratios, resolutions etc. Currently, the driver assumes only one sampling rate but this is only used in the #define n_bytes_per_sample which is used to determine the amount of bytes to be copied. If you want the more flexible version this needs to be generalized in the driver...I did not do this so far because at some point I had to finish a version ;)Am 13.03.21, 09:53 schrieb HiFiPhile _**@**_.***>: Hi @PanRe, I haven't look through the latest audio driver changes, have you implemented interface alternative settings ? For example one for 16 bit, another for 32 bit, and last one for native DSD ? —You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.
> 
> Great ! I must have been confused by CFG_TUD_AUDIO_N_BYTES_PER_SAMPLE_TX

I gave that a bit of thought and i would like to implement this generalization right away. (Note, that the CFG_TUD_AUDIO_N_BYTES_PER_SAMPLE_XX are only used for PCM en/decoding using support FIFOs. So you can do all on your own, support FIFOs are not mandatory).
There is one design question, what do you think is the best way to save the information of the bytes_per_sample? This is optional. Within C, i can not think right now of a nice way to include that into the control struct (i don't know how the user should set it). So i thought of an external array `uint8_t n_bytes_per_sample[CFG_TUD_AUDIO_N_AS_INT]`, which is very easy to implement, yet the user must be aware that such an array needs to be defined by him...
What do you think?

Reply:
or maybe you can set it in one of the callbacks ?

Reply:
Mh i don't know, wouldn't that might require even more in-depth knowledge of the driver?

Reply:
If I remember correctly the maximum CFG_TUD_AUDIO_N_AS_INT is 3. I think it's possible to get n_bytes_per_sample[] from the defines.

For example if `CFG_TUD_AUDIO_N_AS_INT = 2`, `CFG_TUD_AUDIO_N_BYTES_PER_SAMPLE_XX_AS_1` and `CFG_TUD_AUDIO_N_BYTES_PER_SAMPLE_XX_AS_2` should be defined if support fifo are used. 

Element size of fifo could be fixed to the greater one.

Reply:
I just realized that there is a bit more to generalize, the driver should be able to support multiple audio functions. Yet, in C it is a bit difficult to allow such a flexibility e.g. if you have two audio functions, one has no EPs at all (just for controlling of attached audio equipment) and one for streaming with EPs and support FIFOs. (I don't know when such a setup would make sense but USB allows for it ^^). Right now, the first audio function would also setup support FIFOs without needing them. So, there is more to do ^^. I will handle this before this PR is merged because i want the structure of the driver to be fixed to avoid dramatic changes in the future.
@hathach please do not merge the driver for now. What i am about to generilze will not have dramtic impact for this review but i want to have this done once and for all ;)

Reply:
@PanRe got it, it takes me time to review the PR anyway. I am still catching up with your work :)

Reply:
I am currently a bit puzzeled how to design the driver in such a way that it can be generalized conviniently for multiple audio functions. Especially the buffers are a bit tricky. Picture the following:
- two audio driver functions i.e. `CFG_TUD_AUDIO = 2`
- none uses support FIFOs to keep this easy
- the first uses EP IN but needs an additional linear buffer because `usbd_edpt_iso_xfer()` is not available for the target MCU
- the second does not use EP's at all but only controls audio equipment

Currently, the driver would setup buffers for both audio driver functions but the second one does not need any. This is a waste of space. Now, i would usually define pointers to buffers and let the user set buffers up, where the buffers are handed over within an init function. This way only pointers for buffers are setup within the audio driver which for the second driver are not necessary but are not such a big waste of space. However, `audiod_init(void)` does not accept any arguments, so this approach is not possible. The next possibility would be to define extern variables i.e. an array of pointers to buffers, where the user is responsible for setting it up correctly. This requires the user to know when linear buffers are required or not which is not user-friendly.

Is there any method in C such that for a given array of sizes e.g. `const uint8_t[] = {128, 0}` arrays are created of non-equal length? Non-equal length buffers are usually handled by an array of pointers pointing to the buffers. But how to setup those automatically at compile time? Any ideas?

Reply:
Is it possible to add a cb in audiod_init, where we pass all parameters by a struct ?

Reply:
@PanRe there is no need to support multiple audio functions right with  this PR. We can do it later on, Sometimes, a cool-down period is what is needed for a brilliant idea. I have switched back and forth between works. Although It takes a bit of time when revising an unresolved issue, we will have a fresh look at it and could resolve it in a much better/cleaner way (done it multiple times). Just leave it to solve later on if needed. There is already lots of works in this PR :)

Reply:
True, but i really would like to come up with an interface or a way to initialize the audio driver which does not need to be changed afterwards. The users would not appreciate this i guess ^^. But if it is too complicated i would leave it as it is.

I have a pretty fancy (but not too dramatic) idea in my mind. All we need would be a way to setup arrays of different sizes by use of the C preprocessor! Given the following declared list
`
#define AUDIO_EP_IN_SZ_LIST (11, 22)     // Size in bytes, must contain CFG_TUD_AUDIO elements
`
the result of the preprocessor should look like this (where the names could be an argument of the macro)
```
uint8_t audio_ep_in_sw_buffer_1[11];
uint8_t audio_ep_in_sw_buffer_2[22];

ep_in_sw_buffers[CFG_TUD_AUDIO] =
{
    {audio_ep_in_sw_buffer_1},
    {audio_ep_in_sw_buffer_2},
};
```
This is an array of arrays holding different sizes. Since the lengths of the sub-arrays must be known somehow, people came up with a very nice idea to group the array and its corresponding length by use of a struct as done [here](https://forum.arduino.cc/index.php?topic=580090.0)
```
#define NUMELEMENTS(x) (sizeof(x) / sizeof(x[0]))

int pin1[] = {1, 2};
int pin2[] = {9, 8, 7, 6, 5};
int pin3[] = {3, 4};

struct PINS
{
  int *pins;
  uint16_t numElements;
};

PINS pinMatrix[] =
{
  {pin1, NUMELEMENTS(pin1)},
  {pin2, NUMELEMENTS(pin2)},
  {pin3, NUMELEMENTS(pin3)},
};
```
The only tricky part is now to setup a preprocessor macro to generate the above array of arrays from the given defined list. Something related like that was done [here](https://stackoverflow.com/questions/12824703/array-format-for-define-c-preprocessor) but for now i did not find a final example of how the preprocessor would produce the wanted stuff. We would need something like a loop. I know, preprocessors don't posses loops usually, but it would be sufficient if we would come up with something which would setup 10 such arrays at most.

This would be pretty cool and useful anyway to have something which sets up an array of arrays of different size from a list and could be included within `tusb_common.h` file. However, i am not experienced with the preprocessor stuff. Do you know how to do that?

Reply:
Personally I don't like cascade too many macros, it makes debug and understanding difficult.
How about support a max number of audio functions, for example 3, I believe nobody will use more functions. (Also USB descriptor length limit)
Then:
```
uint8_t audio_ep_in_sw_buffer_1[CFG_TUD_AUDIO_FUNC_1_EP_IN_SZ];
#if CFG_TUD_AUDIO > 1
uint8_t audio_ep_in_sw_buffer_2[CFG_TUD_AUDIO_FUNC_2_EP_IN_SZ];
#endif
#if CFG_TUD_AUDIO > 2
uint8_t audio_ep_in_sw_buffer_3[CFG_TUD_AUDIO_FUNC_3_EP_IN_SZ];
#endif

ep_in_sw_buffers[CFG_TUD_AUDIO] =
{
    {audio_ep_in_sw_buffer_1},
#if CFG_TUD_AUDIO > 1
    {audio_ep_in_sw_buffer_2},
#endif
#if CFG_TUD_AUDIO > 2
    {audio_ep_in_sw_buffer_3},
#endif
};

Reply:
That is actually a very nice idea! I think i go with that! :)